docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,235 @@
+
+* Controllers: when creating child objects, controllers need to fill up the their ObjectMeta.ParentReferences field. Objects that don’t have a parent should have the namespace object as the parent.
+
+# Design 2. Finalizer framework + GC
+
+There are two parts in this design. 2A is about the finalizer framework; 2B is about how to use the framework to enforce the orphaning policy.
+
+## 2A. Finalizer framework
+
+## API changes
+
+```
+type ObjectMeta struct {
+	…
+	Finalizers map[string]string
+}
+```
+
+**ObjectMeta.Finalizers**: the map key is an ad hoc, explanatory name(e.g., "orphan"), the value is an identifier for the responsible component that will remove the finalizer from the map. If the ObjectMeta.DeletionTimestamp is non-nil, entry in the map can only be removed.
+
+## New components
+
+* Finalizer:
+  * Like a controller, a finalizer is always running.
+  * A third party can develop their own finalizer. A finalizer doesn't need to be registered with the API server.
+  * Watches for update events that meet two conditions:
+    1. the updated objects has the identifier of the finalizer in ObjectMeta.Finalizers;
+    2. ObjectMeta.DeletionTimestamp is updated from nil to non-nil.
+  * Applies the finalizing logic to the object in the update event.
+  * Removes itself from ObjectMeta.Finalizers after the finalizing logic is completed.
+  * The API server deletes the object after the last finalizer removes itself from the ObjectMeta.Finalizers field.
+  * Because it's possible for the finalizing logic to be applied multiple times, the finalizing logic has to be idempotent.
+  * If a finalizer fails to act in a timely manner, users can manually remove the finalizer from ObjectMeta.Finalizers by a regular kubectl command.
+
+## Changes to existing components
+
+* API server:
+  * Deletion handler:
+    * If the `ObjectMeta.Finalizers` of the object being deleted is non-empty, then updates the DeletionTimestamp, but does not delete the object.
+    * If the `ObjectMeta.Finalizers` is empty and the [options.GracePeriod](../../pkg/registry/generic/etcd/etcd.go#L405) is zero, then deletes the object. If the options.GracePeriod is non-zero, then just updates the DeletionTimestamp.
+  * Update handler:
+    * If the update removes the last finalizer, and the DeletionTimestamp is non-nil, and the DeletionGracePeriodSeconds is zero, then deletes the object from the registry.
+    * If the update removes the last finalizer, and the DeletionTimestamp is non-nil, but the DeletionGracePeriodSeconds is non-zero, then just updates the object.
+
+## 2B. Enforcing orphaning policty with a finalizer
+
+## API changes
+
+```
+type ObjectMeta struct {
+	...
+	ParentReferences []ObjectReference
+}
+```
+
+**ObjectMeta.ParentReferences**: same as in [design #1](#api-changes).
```

Comment:
> It sounds like we allow references to both namespace and non-namespaced resources.

We disallow cross-namespace reference. It makes sense to allow a namespaced object to have a root-scoped object as a parent, I'm not sure if we should allow the reverse case.

> At minimum, I think we need a mechanism to assert what Kinds can reasonably be parents of other Kinds.

I agree. Maybe we can hard-code the rules as validation rules.